pandas/core/generic.py

```diff
@@ -6115,20 +6115,25 @@ def fillna(
                 )
 
             elif isinstance(value, (dict, ABCSeries)):
+                temp_data = self if inplace else self.copy()
+
                 if axis == 1:
-                    raise NotImplementedError(
-                        "Currently only can fill "
-                        "with dict/Series column "
-                        "by column"
-                    )
+                    for i, item in enumerate(temp_data.items()):
+                        label, content = item
```

Comment:
this doesn't make sense with the axis here; you are updating the same column whether axis==0 or 1

Reply:
@jreback 
Yes. but, filled value is different whether axis==0 or 1. And 'downcast' works properly when execute column-based.